[SPARK-17919] Make timeout to RBackend configurable in SparkR

falaki opened the pull request:
## What changes were proposed in this pull request?

This patch makes RBackend connection timeout configurable by user.
## How was this patch tested?

N/A

falaki committed:
Making the timeout of connection to RBackend configurable

SparkQA commented:
**[Test build #66915 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66915/consoleFull)** for PR 15471 at commit [`9dde457`](https://github.com/apache/spark/commit/9dde45721288175a26d5a6e0d0dc42f876e247af).

SparkQA commented:
**[Test build #66915 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66915/consoleFull)** for PR 15471 at commit [`9dde457`](https://github.com/apache/spark/commit/9dde45721288175a26d5a6e0d0dc42f876e247af).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66915/
Test PASSed.

falaki committed:
Setting configurable timeout on monitor and worker connections as well

SparkQA commented:
**[Test build #66928 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66928/consoleFull)** for PR 15471 at commit [`d27233c`](https://github.com/apache/spark/commit/d27233ce454c6ab4d4b208d3d1cffde69789f07a).

SparkQA commented:
**[Test build #66928 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66928/consoleFull)** for PR 15471 at commit [`d27233c`](https://github.com/apache/spark/commit/d27233ce454c6ab4d4b208d3d1cffde69789f07a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66928/
Test PASSed.

felixcheung commented:
would it be possible to pull "6000" into a constant so it could be found/changed in one shot?

falaki commented:
@felixcheung that is a good suggestion. I will try to use a single constant.
I changed the label to WIP because something is still timing out the connection in my tests. Maybe very long timeouts are not properly implemented in R? If you noticed I am setting timeout on all the connections that are every opened. 

@shivaram I am going to experiment with a helper thread inside the main netty thread that sends heartbeats (-1 as result) and then `invokeJava()` would try reading again if it sees the heartbeat value. What do you think?

shivaram commented:
@falaki Thats an interesting approach to try -- the other thing to try might be to use a separate connection to send back results. i.e. as soon as JVM gets the request it returns `OK` or some such status and then whenever the result is ready the JVM pushes data on the reader socket that is created in R. We might need to try a couple of options here to see which looks best.

BTW the monitorConn is used - its used in cases where the JVM comes up first, say in YARN / spark-submit and the JVM uses that to detect if the R process has crashed.

falaki committed:
Using static file for various hard coded default values

falaki committed:
Setting the socket timeout for netty child sockets

falaki committed:
Sending heartbeats to prevent idle connections and timeouts

falaki commented:
@shivaram thanks for clarification. 
I realized we were not setting socket timeout on Netty socket. So I added that as well.
I also introduced the heartbeat mechanism and tested it locally. Next I am going to test if this works on real workload.

SparkQA commented:
**[Test build #66997 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66997/consoleFull)** for PR 15471 at commit [`630467e`](https://github.com/apache/spark/commit/630467efa177b711b2fef8bbcd185136bbf46b80).

SparkQA commented:
**[Test build #66997 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66997/consoleFull)** for PR 15471 at commit [`630467e`](https://github.com/apache/spark/commit/630467efa177b711b2fef8bbcd185136bbf46b80).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66997/
Test FAILed.

falaki committed:
Sharable BackendHandler

falaki committed:
Style

SparkQA commented:
**[Test build #67001 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67001/consoleFull)** for PR 15471 at commit [`3744623`](https://github.com/apache/spark/commit/37446238a8ff0e80775c0e0227945cebc6707bc0).

falaki committed:
Setting ReadTimeoutHandler

SparkQA commented:
**[Test build #67005 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67005/consoleFull)** for PR 15471 at commit [`6f15a15`](https://github.com/apache/spark/commit/6f15a1541f01429ae19237252c600b108722ecb4).

SparkQA commented:
**[Test build #67001 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67001/consoleFull)** for PR 15471 at commit [`3744623`](https://github.com/apache/spark/commit/37446238a8ff0e80775c0e0227945cebc6707bc0).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67001/
Test FAILed.

SparkQA commented:
**[Test build #67005 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67005/consoleFull)** for PR 15471 at commit [`6f15a15`](https://github.com/apache/spark/commit/6f15a1541f01429ae19237252c600b108722ecb4).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67005/
Test FAILed.

falaki commented:
@shivaram this worked on my stress tests. The question is how to unit test this?

yhuai commented:
test this please

SparkQA commented:
**[Test build #67079 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67079/consoleFull)** for PR 15471 at commit [`6f15a15`](https://github.com/apache/spark/commit/6f15a1541f01429ae19237252c600b108722ecb4).

shivaram commented:
@falaki Do we know if the tests timeout are due to this change ? Or are they unrelated ?

SparkQA commented:
**[Test build #67079 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67079/consoleFull)** for PR 15471 at commit [`6f15a15`](https://github.com/apache/spark/commit/6f15a1541f01429ae19237252c600b108722ecb4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67079/
Test FAILed.

falaki commented:
@shivaram I think they are unrelated. Can you trigger another test?

shivaram commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #67091 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67091/consoleFull)** for PR 15471 at commit [`6f15a15`](https://github.com/apache/spark/commit/6f15a1541f01429ae19237252c600b108722ecb4).

SparkQA commented:
**[Test build #67091 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67091/consoleFull)** for PR 15471 at commit [`6f15a15`](https://github.com/apache/spark/commit/6f15a1541f01429ae19237252c600b108722ecb4).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67091/
Test FAILed.

shivaram commented:
@falaki looks like the SparkR MLlib unit tests are timing out on Jenkins. Do they pass on your machine ?

falaki committed:
Use +1 for heartbeat

falaki commented:
@shivaram it was indeed my fault. I did not run local tests after I added the heartbeat. I am now using +1 for heartbeat.

SparkQA commented:
**[Test build #67153 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67153/consoleFull)** for PR 15471 at commit [`fcc3376`](https://github.com/apache/spark/commit/fcc33764259b75f566adf9573cd503cac66d32e0).

SparkQA commented:
**[Test build #67153 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67153/consoleFull)** for PR 15471 at commit [`fcc3376`](https://github.com/apache/spark/commit/fcc33764259b75f566adf9573cd503cac66d32e0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67153/
Test PASSed.

felixcheung commented on core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala:
```diff
@@ -83,7 +86,29 @@ private[r] class RBackendHandler(server: RBackend)
+      }
+      val conf = new SparkConf()
+      val heartBeatInterval = conf.getInt(
+        "spark.r.heartBeatInterval", SparkRDefaults.DEFAULT_HEARTBEAT_INTERVAL)
```
should this be documented too?

falaki replied:
Done.

felixcheung commented on core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala:
```diff
@@ -83,7 +86,29 @@ private[r] class RBackendHandler(server: RBackend)
       }
     } else {
+      // To avoid timeouts when reading results in SparkR driver, we will be regularly sending
+      // heartbeat responses. We use special character -1 to signal the client that backend is
```
`-1` -> `+1`?

felixcheung commented on R/pkg/R/backend.R:
```diff
@@ -108,13 +108,27 @@ invokeJava <- function(isStatic, objId, methodName, ...) {
   returnStatus <- readInt(conn)
+  handleErrors(returnStatus, conn)
+
+  # Backend will send -1 as keep alive value to prevent various connection timeouts
```
`-1` -> `+1`?

felixcheung commented on core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala:
```diff
@@ -95,9 +120,15 @@ private[r] class RBackendHandler(server: RBackend)
+    cause match {
+      case timeout: ReadTimeoutException =>
+        // Do nothing. We don't want to timeout on read
+        logInfo("Ignoring read timeout in RBackendHandler")
```
logWarning?

felixcheung reviewed (changes requested):
Looks good aside from minor comments.
This goes to 2.1.0?

falaki committed:
Addressed comments

falaki commented:
Thanks @felixcheung addressed your comments.

SparkQA commented:
**[Test build #67287 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67287/consoleFull)** for PR 15471 at commit [`749c8b0`](https://github.com/apache/spark/commit/749c8b0967993d5b8e08dc84904cc46a1ce52923).

felixcheung commented:
LGTM

SparkQA commented:
**[Test build #67287 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67287/consoleFull)** for PR 15471 at commit [`749c8b0`](https://github.com/apache/spark/commit/749c8b0967993d5b8e08dc84904cc46a1ce52923).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67287/
Test PASSed.

shivaram commented:
@falaki @felixcheung Since this is a big change I'd like to also take a look at this once - Will try to get to it tonight.

felixcheung commented:
sure. I think the target is 2.0.2 - it will be good to review this more closely.

falaki commented:
Thanks @shivaram. I ran a real workload consisting of long running parallel simulations that took about 3.5 hours. I also tested it by calling `Sys.sleep()` inside workers with dapply and spark.lapply. Would be good to torture it in new ways. 

Also would be interesting to think about ways of unit-testing this.

shivaram commented on R/pkg/inst/worker/worker.R:
```diff
@@ -90,6 +90,7 @@ bootTime <- currentTimeSecs()
 bootElap <- elapsedSecs()
 
 rLibDir <- Sys.getenv("SPARKR_RLIBDIR")
+connectionTimeout <- Sys.getenv("SPARKR_BACKEND_CONNECTION_TIMEOUT")
```
Default value here as well

shivaram commented on core/src/main/scala/org/apache/spark/api/r/RBackend.scala:
```diff
@@ -110,6 +115,11 @@ private[spark] object RBackend extends Logging {
+      // Connection timeout is set by socket client. To make it configurable we will pass the
+      // timeout value to client inside the temp file
+      val conf = new SparkConf()
+      val backendConnectionTimeout = conf.getInt(
```
Are we sure SparkConf has been initialized successfully at this point ? Or to to put it another way, in which cases does this code path get called from ? Is this in the `spark-submit` case or the shell etc. ?

falaki replied:
This is for `spark-submit`. Basically the JVM starts before the R process. As a result the only way for R process to get these configuration parameters from the JVM. In this case, `RBackend` sets the environment variables based on configs.

For the other mode where JVM is started after the R process, we are sending this timeout value through the TCP connection.

At least that is my current understanding of how deploy modes work. In our production environment we launch the R process from the JVM.

shivaram commented on core/src/main/scala/org/apache/spark/api/r/RRunner.scala:
```diff
@@ -333,6 +333,8 @@ private[r] object RRunner {
     var rCommand = sparkConf.get("spark.sparkr.r.command", "Rscript")
     rCommand = sparkConf.get("spark.r.command", rCommand)
 
+    val rConnectionTimeout = SparkEnv.get.conf.getInt(
```
can we just use `sparkConf.get` similar to the line above ?

shivaram commented on core/src/main/scala/org/apache/spark/api/r/RBackendHandler.scala:
```diff
@@ -83,7 +86,29 @@ private[r] class RBackendHandler(server: RBackend)
+      // To avoid timeouts when reading results in SparkR driver, we will be regularly sending
+      // heartbeat responses. We use special code +1 to signal the client that backend is
+      // alive and it should continue blocking for result.
+      val execService = ThreadUtils.newDaemonSingleThreadScheduledExecutor("SparkRKeepAliveThread")
```
I'm not sure how expensive it is to create and destroy an executor service each time. Can we just schedule at fixed rate when we get the request and then cancel the schedule at the end of the request ?

falaki replied:
I was not sure about this either. I used this method based on advice from @zsxwing.

shivaram replied:
Hmm - my question on whether we can reuse this still stands. @zsxwing do you think thats possible ?

shivaram replied:
I took a closer look at this and (a) it looks like the executor just calls cancel on the tasks during shutdown, so that part of the behavior is the same as calling cancel on the task we have [1]. But you are right that if we wanted to wait for termination we'd need to do some extra work.  We could use the `get` call but its unclear what the semantics of that are. It might be more easier to just setup a semaphore or mutex that is shared by the runnable and the outside thread.

But overall it looks like thread pool creation is only around 100 micro-seconds [2] and I also benchmarked this locally.

[1] http://www.docjar.com/html/api/java/util/concurrent/ScheduledThreadPoolExecutor.java.html line 367
[2] http://stackoverflow.com/a/5483467/4577954

shivaram reviewed (commented):
Thanks @falaki -- This is a very useful change. I just did a pass of comments.

falaki committed:
Addressed comments

falaki commented:
@shivaram sorry for delay getting back to this. Please take another look.

SparkQA commented:
**[Test build #67602 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67602/consoleFull)** for PR 15471 at commit [`666b609`](https://github.com/apache/spark/commit/666b60947b46ca013b9a0a7dd014f2e6580ceb50).

SparkQA commented:
**[Test build #67602 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67602/consoleFull)** for PR 15471 at commit [`666b609`](https://github.com/apache/spark/commit/666b60947b46ca013b9a0a7dd014f2e6580ceb50).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67602/
Test PASSed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #67661 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67661/consoleFull)** for PR 15471 at commit [`666b609`](https://github.com/apache/spark/commit/666b60947b46ca013b9a0a7dd014f2e6580ceb50).

SparkQA commented:
**[Test build #67661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67661/consoleFull)** for PR 15471 at commit [`666b609`](https://github.com/apache/spark/commit/666b60947b46ca013b9a0a7dd014f2e6580ceb50).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67661/
Test FAILed.

falaki commented:
retest this please

SparkQA commented:
**[Test build #67667 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67667/consoleFull)** for PR 15471 at commit [`666b609`](https://github.com/apache/spark/commit/666b60947b46ca013b9a0a7dd014f2e6580ceb50).

SparkQA commented:
**[Test build #67667 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67667/consoleFull)** for PR 15471 at commit [`666b609`](https://github.com/apache/spark/commit/666b60947b46ca013b9a0a7dd014f2e6580ceb50).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67667/
Test PASSed.

falaki commented:
@shivaram is there a chance this makes it to the 2.0.2 release?

shivaram commented:
Taking another look now

shivaram commented:
@falaki The code change looks pretty good to me, but I'm still a bit worried about introducing a big change in a minor release. Can we have this disabled by default and flip the flag only in the master branch ?

@HyukjinKwon Is there any way to retrigger the AppVeyor build ?

HyukjinKwon commented:
Build started: [SparkR] `ALL` [![PR-15471](https://ci.appveyor.com/api/projects/status/github/spark-test/spark?branch=1175779D-A053-45AF-BC6C-EA34931CFC37&svg=true)](https://ci.appveyor.com/project/spark-test/spark/branch/1175779D-A053-45AF-BC6C-EA34931CFC37)
Diff: https://github.com/apache/spark/compare/master...spark-test:1175779D-A053-45AF-BC6C-EA34931CFC37

@shivaram I first thought committers can access to AppVeyor account but it seems not. We should go to the Web UI and click the rebuild button.. So, I just made a bunch of scripts to launch a build via @spark-test account in such cases. Please cc me (treat me like a bot) then I will leave such comments above until we find a good way to do it.

falaki commented:
@shivaram that is fine. We can merge it to 2.1 (or whatever the next major release is going to be).

shivaram commented:
Thanks @HyukjinKwon - The appveyor tests seem to pass as well

Change LGTM to merge to master. @felixcheung Any other comments ?

felixcheung commented:
LGTM.

felixcheung commented:
merged to master.

shivaram commented on R/pkg/R/backend.R:
```diff
@@ -108,13 +108,27 @@ invokeJava <- function(isStatic, objId, methodName, ...) {
+
+  # Backend will send -1 as keep alive value to prevent various connection timeouts
+  # on very long running jobs. See spark.r.heartBeatInterval
+  while (returnStatus == 1) {
```
@falaki @felixcheung any thoughts on this ?

felixcheung replied:
+1 I think it's a good idea to avoid infinite loop in general.
how is toree calling this?
could you open a JIRA?
